Create an tbf parsing crate #1

mihai-negru opened the pull request:
### Pull Request Overview

The idea of this pull request is to set the foundation for building a rust version of `tockloader`. The arguments for this are:
- pip is not very portable when it comes to installing commands, it sometimes works just with `pip`, otherwise it needs `pip --user` or `sudo`.
- unifying the tools and libraries used for tock, so that we do not use two programming languages


#### Created a new workspace
* Workspace at the root level
* Two containing members elf2tab and tbfparser

#### Modifications in the elf2tab
* Removed the header.rs file and added to a new created crate
* Removed the utils.rs file because it can be found at the tbfparser library
* Modified Cargo.toml dependencies by adding the new created crate, the path is specified relative to the root project folder.
* Replaced the header file to tbfparser package.

#### Creating the new library crate **tbfparser**
* Created a **library** cargo crate.
* Moved header.rs to the lib.rs.
* Copied utils.rs in the src folder of the current crate.

### Testing Strategy
This pull request was tested by compiling.
Copiled the project in the **workspace** using *cargo build*, resulting in elf2tab elf file.

### TODO or Help Wanted
N/A

### Documentation Updated
Added a new *subsubsection* in the **elf2tabDetails** to describe the project's structure. (workflow)

alexandruradovici commented:
The tbf parser should be in its own crate, different from elf2tab. I would create a wrokspace in the root folder (https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html) and place the two crates, elf2tab and tbfparser, in the workspace. Adapt the readme to this.

mihai-negru commented:
> The tbf parser should be in its own crate, different from elf2tab. I would create a wrokspace in the root folder (https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html) and place the two crates, elf2tab and tbfparser, in the workspace. Adapt the readme to this.

I have created the workspace and moved both crates into the workspace. Also updated the readme file for compilation instructions.

alexandruradovici reviewed (changes requested):
I would move the workspace to the root.

mihai-negru commented:
> I would move the workspace to the root.

I have moved the workspace to the root level and removed the utils.rs from the elf2tab crate, to remove code duplication (utils.rs can be also found in the tbfparser crate as a public module). However I do not find something in the README file to modify, should I create a new section to write explicity that the repository contains one workspace with two crates?

P.S: I have also edited the PR overview

alexandruradovici commented on tbfparser/src/lib.rs:
```diff
@@ -1,10 +1,11 @@
+use crate::util::amount_alignment_needed;
 
 #[repr(u16)]
 #[derive(Clone, Copy, Debug)]
```
Is the allow dead code still required here? You are exporting this structure, so the it should not be necessary.

mihai-negru replied:
No, it is not required. Commited a new change to remove the allow_dead

alexandruradovici commented on tbfparser/src/util.rs:
```diff
@@ -9,6 +9,7 @@ pub fn align_to(value: u32, box_size: u32) -> u32 {
 }
 
 /// Takes a value and rounds it down to be aligned % box_size
+#[allow(dead_code)]
```
Is this still required?

alexandruradovici commented:
> However I do not find something in the README file to modify, should I create a new section to write explicity that the repository contains one workspace with two crates?
> 
I think this is a good solution.

alexandruradovici commented on README.md:
```diff
@@ -177,6 +177,14 @@ read/write permissions on existing data.
+#### Workflow
+
+The project is encapsulated in a rust workspace, containing two cargo crates:
+* *elf2tab* crate - containing the main.rs file.
```
```suggestion
* *elf2tab* crate - containing the source code for the `elf2tab` binary.
```

mihai-negru replied:
Done

alexandruradovici commented on tbfparser/src/lib.rs:
```diff
@@ -359,7 +356,7 @@ impl TbfHeader {
             header_length += mem::size_of::<TbfHeaderTlv>() + package_name.len();
             // How much padding is needed to ensure we are aligned to 4?
-            let pad = amount_alignment_needed(header_length as u32, 4);
+            let pad = util::amount_alignment_needed(header_length as u32, 4);
```
```suggestion
            let pad = amount_alignment_needed(header_length as u32, 4);
```

Add `use util::amount_alignment_needed` at the top of the file.

mihai-negru replied:
Done

mihai-negru commented:
> > However I do not find something in the README file to modify, should I create a new section to write explicity that the repository contains one workspace with two crates?
> 
> I think this is a good solution.

I created a subsubsection under the **Details** section from the README and described very shortly how the project is structured.

alexandruradovici commented on tbfparser/src/lib.rs:
```diff
@@ -682,7 +679,7 @@ impl TbfHeader {
         util::do_pad(
             &mut header_buf,
-            amount_alignment_needed(current_length as u32, 4) as usize,
+            util::amount_alignment_needed(current_length as u32, 4) as usize,
```
```suggestion
            amount_alignment_needed(current_length as u32, 4) as usize,
```
Add `use util::amount_alignment_needed` at the top of the file.

mihai-negru replied:
Done

alexandruradovici commented on Cargo.toml:
```diff
@@ -1,23 +1,6 @@
+members = [
+    "elf2tab",
+    "tbfparser",
+]
```
Add an empty new line.

mihai-negru replied:
Done

alexandruradovici commented on .gitignore:
```diff
@@ -1,2 +1 @@
-target
-
+target
```
add and empty new line

mihai-negru replied:
Done

alexandruradovici replied:
The file still seems to miss the newline.

mihai-negru replied:
Sorry, added now the endling newline

alexandruradovici commented on elf2tab/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+sha2 = "0.10.2"
+ring = "0.16.20"
+rsa-der = "0.3.0"
+tbfparser = { path = "../tbfparser" }
```
Add an empty new line.

mihai-negru replied:
Done

alexandruradovici commented on elf2tab/src/lib.rs:
```diff
@@ -0,0 +1,2 @@
+pub mod cmdline;
+pub mod convert;
```
Please add a newline here.

mihai-negru replied:
Done

alexandruradovici commented:
Formatting fails, please run `cargo fmt` on all the crates.

mihai-negru commented:
> Formatting fails, please run `cargo fmt` on all the crates.

Running cargo fmt --check on my machine seems to work
![image](https://user-images.githubusercontent.com/86746059/206659940-d672e430-4670-4006-842f-3ee9c1994b24.png)

alexandruradovici commented:
Please open a pull request to the upstream repository and post here a link.

mihai-negru commented:
Sorry for the question, how can I open the pull request from a fork that is forked from the main repository? Or it should be merged to this upstream firstly and then to the main repository. Thanks for the answer

alexandruradovici commented:
Open a new pull request and select the upstream (original) repository. You should have a dropdown there.

mihai-negru commented:
Link to the pull request : https://github.com/tock/elf2tab/pull/62

mihai-negru commented:
Hello, sorry for the question. I wanted to ask if is anything more to do on this pull request, I do not seem to understand what I should do next by reading the bradjc's comment from the main stream. Thanks in advice!

alexandruradovici commented:
What Brad was suggesting is that you take a look at [tock-tbf](https://github.com/tock/tock/tree/master/libraries/tock-tbf) crate and see if we can merge this. This may or not be possible, as tock-tbf does not depend on the std crate (`#[no-std]`) and is not able to use memory allocation.

alexandruradovici commented:
Any update here?

mihai-negru commented:
> Any update here?

I have looked over the tock-tbf and revised the tbfparser and for me, it seems impossible to modify the tbfparser code in order no to use the standard library (#no-std), as suggested bradjc creating a separate repository just for the tbfparser would make the deal.

Matrix22 commented:
> Any update here?

I have looked over the tock-tbf and revised the tbfparser and for me, it seems impossible to modify the tbfparser code in order no to use the standard library (#no-std), as suggested bradjc creating a separate repository just for the tbfparser would make the deal.

alexandruradovici commented:
One approach would be to use the data types from the tbf crate defined in tock instead of redefining them. What about adding the crate in the tock repository, next to the tbf parser?

mihai-negru commented:
I closed the last pull request and I will work on the new idea that you have suggested, on using the types defined in the tock/libraries/tbfparser

> One approach would be to use the data types from the tbf crate defined in tock instead of redefining them. What about adding the crate in the tock repository, next to the tbf parser?
